### PR TITLE
feat: Exclude certain endpoints from media type validation

### PR DESCRIPTION
**Pull request by estahn**
- Fixes justinrainbow/json-schema#543 and justinrainbow/json-schema#540

### CODE REVIEW

**Comment by estahn** [comment]
@erayd I'm flexible on the naming of the `attribute` and `method` name. I just didn't have better ideas. Let me know your thoughts.

File: src/JsonSchema/Uri/UriRetriever.php [hunk 1]
```diff
@@ -32,6 +32,14 @@ class UriRetriever implements BaseUriRetrieverInterface
         '|^https?://json-schema.org/draft-(0[34])/schema#?|' => 'package://dist/schema/json-schema-draft-$1.json'
     );
 
+    /**
+     * @var array A blacklist for media type ceheck exclusion
```
Comment (erayd) [inline src/JsonSchema/Uri/UriRetriever.php]: Typo in comment ('ceheck')

File: src/JsonSchema/Uri/UriRetriever.php [hunk 2]
```diff
@@ -32,6 +32,14 @@ class UriRetriever implements BaseUriRetrieverInterface
...
+     */
+    protected $mediaTypeBlacklist = array(
```
Comment (erayd) [inline src/JsonSchema/Uri/UriRetriever.php]: More descriptive name please - perhaps something like 'allowInvalidContentType'.

File: src/JsonSchema/Uri/UriRetriever.php [hunk 3]
```diff
@@ -44,6 +52,16 @@ class UriRetriever implements BaseUriRetrieverInterface
      */
     private $schemaCache = array();
 
+    /**
+     * Adds an endpoint to the media type validation blacklist
+     *
+     * @param string $endpoint
+     */
+    public function addBlacklistedEndpoint($endpoint)
```
Comment (erayd) [inline src/JsonSchema/Uri/UriRetriever.php]: Again with the naming. I don't feel that this is a blacklist - it's a list of exceptions.

File: src/JsonSchema/Uri/UriRetriever.php [hunk 4]
```diff
@@ -65,9 +83,10 @@ public function confirmMediaType($uriRetriever, $uri)
             return;
         }
 
-        if (substr($uri, 0, 23) == 'http://json-schema.org/') {
-            //HACK; they deliver broken content types
-            return true;
+        for ($i = 0, $iMax = count($this->mediaTypeBlacklist); $i < $iMax; $i++) {
```
Comment (erayd) [inline src/JsonSchema/Uri/UriRetriever.php]: Refactor to use `foreach` please.

File: src/JsonSchema/Uri/UriRetriever.php (hunk 1 above)
Comment (bighappyface) [inline src/JsonSchema/Uri/UriRetriever.php]: spelling

Comment (estahn) [reply src/JsonSchema/Uri/UriRetriever.php]: fixed.

File: src/JsonSchema/Uri/UriRetriever.php [hunk 5]
```diff
@@ -65,9 +83,10 @@ public function confirmMediaType($uriRetriever, $uri)
...
+            if (stripos($uri, $this->mediaTypeBlacklist[$i]) === 0) {
```
Comment (erayd) [inline src/JsonSchema/Uri/UriRetriever.php]: Should be case-sensitive.

File: tests/Uri/UriRetrieverTest.php [hunk 6]
```diff
@@ -339,6 +339,28 @@ public function testJsonSchemaOrgMediaTypeHack()
...
+    /**
+     * @expectedException \JsonSchema\Exception\InvalidSchemaMediaTypeException
+     */
+    public function testJsonSchemaOrgMediaTypeBlacklistUnknown()
+    {
+        $mock = $this->getMock('JsonSchema\Uri\UriRetriever', array('getContentType'));
+        $mock->method('getContentType')->willReturn('Application/X-Fake-Type');
+        $retriever = new UriRetriever();
+
+        $retriever->confirmMediaType($mock, 'http://iglucentral.com');
```
Comment (erayd) [inline tests/Uri/UriRetrieverTest.php]: Use example.com for example domains please.

**Comment by bighappyface** [comment]
@estahn is _blacklist_ the correct term? Wouldn't a list of allowed endpoints be a whitelist?

File: tests/Uri/UriRetrieverTest.php [hunk 7]
```diff
@@ -339,6 +339,28 @@ public function testJsonSchemaOrgMediaTypeHack()
...
+    }
+
+    public function testJsonSchemaOrgMediaTypeBlacklistAdded()
+    {
+        $mock = $this->getMock('JsonSchema\Uri\UriRetriever', array('getContentType'));
+        $mock->method('getContentType')->willReturn('Application/X-Fake-Type');
+        $retriever = new UriRetriever();
+        $retriever->addBlacklistedEndpoint('http://iglucentral.com');
```
Comment (erayd) [inline tests/Uri/UriRetrieverTest.php]: Use example.com for example domains please.

**Comment by estahn** [comment]
> @estahn is _blacklist_ the correct term? Wouldn't a list of allowed endpoints be a whitelist?

Indeed, but since the list contains excluded endpoints it would be a blacklist. I'm happy to rename it to something else, e.g. `excludedMediaTypeCheckEndpoints`.

File: tests/Uri/UriRetrieverTest.php (hunk 6 above)
Comment (estahn) [reply tests/Uri/UriRetrieverTest.php]: fixed.

File: tests/Uri/UriRetrieverTest.php (hunk 7 above)
Comment (estahn) [reply tests/Uri/UriRetrieverTest.php]: fixed.

File: src/JsonSchema/Uri/UriRetriever.php (hunk 5 above)
Comment (estahn) [reply src/JsonSchema/Uri/UriRetriever.php]: @erayd Can you elaborate? Which part needs to be case-sensitive?

File: src/JsonSchema/Uri/UriRetriever.php (hunk 2 above)
Comment (estahn) [reply src/JsonSchema/Uri/UriRetriever.php]: @erayd As I said I'm not happy with the name either and open for suggestions. `allowInvalidContentType` doesn't match the content either I believe.

How about `allowedInvalidContentTypeEndpoints`?

Comment (erayd) [reply src/JsonSchema/Uri/UriRetriever.php]: That works :+1:

File: src/JsonSchema/Uri/UriRetriever.php (hunk 3 above)
Comment (bighappyface) [reply src/JsonSchema/Uri/UriRetriever.php]: @estahn ☝️

File: src/JsonSchema/Uri/UriRetriever.php (hunk 5 above)
Comment (erayd) [reply src/JsonSchema/Uri/UriRetriever.php]: @estahn Most of a URL is case-sensitive; only the domain and scheme are not. As this is a feature which does prefix matching on a full URL, we should be checking case. So `stripos()` is not appropriate here.

An alternative would be to parse the URL properly and separate out the origin, if you would like to distinguish between the sensitive and insensitive parts.

Comment (estahn) [reply src/JsonSchema/Uri/UriRetriever.php]: @erayd I see. Would you consider something like `strtolower` is ok, given it's more a theoretical issue?

Comment (erayd) [reply src/JsonSchema/Uri/UriRetriever.php]: @estahn No, I do not consider that to be OK. It's not just a theoretical issue; this has practical implications (it's a security problem, and this library is widely used). `strtolower` doesn't solve anything, it just transfers the case-insensitivity to a different place.

Comment (erayd) [reply src/JsonSchema/Uri/UriRetriever.php]: @estahn Yes, using `strpos` is OK :-).

File: src/JsonSchema/Uri/UriRetriever.php (hunk 4 above)
Comment (estahn) [reply src/JsonSchema/Uri/UriRetriever.php]: done

File: tests/Uri/UriRetrieverTest.php [hunk 8]
```diff
@@ -330,7 +330,7 @@ public function testRetrieveSchemaFromPackage()
         $this->assertEquals('454f423bd7edddf0bc77af4130ed9161', md5(json_encode($schema)));
     }
 
-    public function testJsonSchemaOrgMediaTypeHack()
+    public function testJsonSchemaOrgMediaTypeBlacklistDefault()
```
Comment (erayd) [inline tests/Uri/UriRetrieverTest.php]: Naming - not a blacklist.

File: tests/Uri/UriRetrieverTest.php [hunk 9]
```diff
@@ -339,6 +339,28 @@ public function testJsonSchemaOrgMediaTypeHack()
         $this->assertTrue($retriever->confirmMediaType($mock, 'http://json-schema.org/'));
     }
 
+    /**
+     * @expectedException \JsonSchema\Exception\InvalidSchemaMediaTypeException
+     */
+    public function testJsonSchemaOrgMediaTypeBlacklistUnknown()
```
Comment (erayd) [inline tests/Uri/UriRetrieverTest.php]: Naming - not a blacklist, and nothing to do with json-schema.org.

File: tests/Uri/UriRetrieverTest.php [hunk 10]
```diff
@@ -339,6 +339,28 @@ public function testJsonSchemaOrgMediaTypeHack()
...
+    {
+        $mock = $this->getMock('JsonSchema\Uri\UriRetriever', array('getContentType'));
+        $mock->method('getContentType')->willReturn('Application/X-Fake-Type');
+        $retriever = new UriRetriever();
+
+        $retriever->confirmMediaType($mock, 'http://example.com');
+    }
+
+    public function testJsonSchemaOrgMediaTypeBlacklistAdded()
```
Comment (erayd) [inline tests/Uri/UriRetrieverTest.php]: Naming - not a blacklist, and nothing to do with json-schema.org.

File: tests/Uri/UriRetrieverTest.php (hunk 8 above)
Comment (estahn) [reply tests/Uri/UriRetrieverTest.php]: done

File: tests/Uri/UriRetrieverTest.php (hunk 9 above)
Comment (estahn) [reply tests/Uri/UriRetrieverTest.php]: done

**Comment by erayd** [comment]
@bighappyface Your turn :-).

**Pull request merged**